rippled Troubleshooting Docs

mDuo13 opened the pull request:
Introduces a "troubleshooting" sub-category in the Managing the rippled server section and carries over some of the more relevant parts of the old [rippled Troubleshooting Wiki page](https://wiki.ripple.com/Rippled_Troubleshooting). (Most of that page is long out of date, though.)

Explains some common steps for diagnosing problems, the causes of some common error messages, and how to fix them.

**Note:** After #448 is merged, I plan to move the SQLite page size doc from that PR into the troubleshooting sub-category from this PR. Until then, this doc won't mention the SQLite thing. Also, the improved CSS for the "text" style is in that PR, so the error messages may look a little weird if you build this PR directly.

Opening so I can get reviewer feedback, but not intended to be merged until after the TODO's are addressed at minimum.

mDuo13 committed:
Start troubleshooting category

mDuo13 committed:
Move troubleshooting to subcategory w/ new diagnosing stub

mDuo13 committed:
More rippled Troubleshooting content

- Fill out "Diagnosing Problems" page
- Add a couple more cases to the page on why rippled won't start

mDuo13 committed:
Expand troubleshooting docs

mDuo13 committed:
Split off log messages from diagnosing rippled problems

MarkusTeufelberger commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+- **`peers`** - This field indicates how many other servers in the XRP Ledger peer-to-peer network your server is connected to. Healthy servers typically show between 5 and 50 peers, unless explicitly configured to connect only to certain peers.
+
+    - If you have 0 peers, your server may be unable to contact the network, or your system clock may be wrong. (Ripple recommends running an [NTP](http://www.ntp.org/) daemon on all servers to keep their clocks synced.)
```
Having exactly 10 peers can indicate being behind a NATing router and a port forward for the peer port missing.

bachase commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+The response to this command has a lot of information, which is documented along with the [server_info method][].
+For troubleshooting purposes, the most important fields are (from most commonly used to least):
+
+- **`server_state`** - Most of the time, this field should show either `full` or `proposing` depending on whether it is [configured as a validator](run-rippled-as-a-validator.html). The value `connected` means that the server can communicate with the rest of the peer-to-peer network, but it does not yet have enough data to track progress of the shared ledger state. Normally, syncing to the state  ...
```
Should you indicate explicitly that `full` is the expected state for a non-validator, but `proposing` is the expected state for a validator?

gbarr01 replied:
Maybe: 

> - **`server_state`** - This field returns `full` if the server is a non-validator and `proposing` if [configured as a validator](run-rippled-as-a-validator.html).

bachase commented on content/tutorials/manage-the-rippled-server/troubleshooting/understanding-log-messages.md:
````diff
@@ -0,0 +1,139 @@
+2018-Aug-28 22:56:36.181169931 LoadMonitor:WRN Job: AcquisitionDone run: 0ms wait: 11566ms
+```
+
+***TODO: how serious is this?***
````
Should get some other folks to weigh in, but I view multisecond `run:` messages like this one as a bad sign. Rare occurrences are ok, but regularly is another sign of resource issues--either directly or contention with something else on the box. Similar to comments you have [above](https://github.com/ripple/ripple-dev-portal/pull/453/files#diff-8f38126fe6f6261393f95ebbf7022aeaR23)

bachase reviewed (approved):
👍 

Did a first pass read and looks good. I'll try to come back for another pass, but ideally a few others on the team and in the community can review too.

scottschurr commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+If you have having problems with `rippled`, the first step is to collect more information to accurately characterize the problem. From there, it can be easier to figure out a root cause and a fix.
+
+If your server does not start, see [rippled Server Won't Start](server-wont-start.html) for a list of some possible causes and fixes.
```
It might be nice to expand a bit on what you mean by "does not start".  This covers catastrophic errors where the server dies and returns to the command prompt.  Some folks think that "does not start" includes does not sync.

scottschurr commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+- **`server_state`** - Most of the time, this field should show either `full` or `proposing` depending on whether it is [configured as a validator](run-rippled-as-a-validator.html). The value `connected` means that the server can communicate with the rest of the peer-to-peer network, but it does not yet have enough data to track progress of the shared ledger state. Normally, syncing to the state  ...
+
+    - If your server remains in the `connected` state for hours, or returns to the `connected` state after being in the `full` or `proposing` states, that usually indicates that your server cannot keep up with the rest of the network. The most common bottlenecks are disk I/O and network bandwidth.
```
Insufficient RAM can cause this as well.

scottschurr commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+- The `rippled` server is just starting up, or is not running at all. Check the status of the service; if it is running, wait a few seconds and try again.
+- You may need to pass different [parameters to the `rippled` commandline client](commandline-usage.html#client-mode-options) to connect to your server.
+- The `rippled` server may not be configured not to accept JSON-RPC connections.
```
Extra "not"?  "...may not be configured ~not~ to..."

intelliot commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+- **`server_state`** - Most of the time, this field should show either `full` or `proposing` depending on whether it is [configured as a validator](run-rippled-as-a-validator.html). The value `connected` means that the server can communicate with the rest of the peer-to-peer network, but it does not yet have enough data to track progress of the shared ledger state. Normally, syncing to the state  ...
+
+    - If your server remains in the `connected` state for hours, or returns to the `connected` state after being in the `full` or `proposing` states, that usually indicates that your server cannot keep up with the rest of the network. The most common bottlenecks are disk I/O and network bandwidth.
```
If your server seems stuck in the `connected` state or `syncing` state, and you are using SSDs, then you should try switching to NuDB (if you are currently using RocksDB).

1. Stop the server (ctrl+c)
2. Edit `rippled.cfg`, updating the `[node_db]` stanza. Example:
```
[node_db]
type=nudb
path=/var/lib/rippled/db/nudb
open_files=2000
filter_bits=12
cache_mb=256
file_size_mb=8
file_size_mult=2
advisory_delete=0
```

To avoid conflicts, make sure the `path` is different from what you were using for RocksDB.

The `open_files`, `filter_bits`, `cache_mb`, `file_size_mb`, and `file_size_mult` settings do not apply to NuDB, but there's no harm in leaving them; and if they remain, it may be easier to switch back to RocksDB if needed.

mtrippled replied:
This is false. Either rocksdb or nudb should be fine on SSD. In fact, nobody should run rippled without server grade SSD's. RocksDB becomes unusable based on memory consumption after a large amount of history is gathered, however. Though I cannot remember how much offhand.

scottschurr commented on content/tutorials/manage-the-rippled-server/troubleshooting/understanding-log-messages.md:
````diff
@@ -0,0 +1,139 @@
+2018-Aug-28 22:56:22.368499966 LedgerConsensus:WRN {"accepted":true,"account_hash":"89A821400087101F1BF2D2B912C6A9F2788CC715590E8FA5710F2D10BF5E3C03","close_flags":0,"close_time":588812130,"close_time_human":"2018-Aug-28 22:55:30.000000000","close_time_resolution":30,"closed":true,"hash":"96A8DF9ECF5E9D087BAE9DDDE38C197D3C1C6FB842C7BB770F8929E56CC71661","ledger_hash":"96A8DF9ECF5E9D087BAE9DDDE38C ...
+```
+
+During the first 5 to 15 minutes after the server starts up, it is normal for it to be out of sync with the rest of the network and print messages such as these. If the server writes these messages long after starting up, it could indicate a problem. Common causes include unreliable network connections and insufficient hardware specs.
````
Can also be caused by background processes sporadically waking up and getting in the way of the server.  Look for things like virus scanners, backups, or processes for cleaning databases.

scottschurr reviewed (approved):
What useful docs!  Thanks a bunch!  I left a couple of comments that may be useful, but entirely at your discretion.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+# Diagnosing Problems with rippled
+
+If you have having problems with `rippled`, the first step is to collect more information to accurately characterize the problem. From there, it can be easier to figure out a root cause and a fix.
```
```
< If you have having problems ... the problem.
---
> If you are having problems ... the issue.
```

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
````diff
@@ -0,0 +1,74 @@
+rippled server_info
+```
+
+The response to this command has a lot of information, which is documented along with the [server_info method][].
````
Is this deliberate? 

> [server_info method][].

There are few of these throughout the docs.

mDuo13 replied:
Yes. The `[x method][]` links are supplied by the `rippled-api-links.md` snippet, which is only included if you use Dactyl to build the docs instead of viewing the Markdown files in, say, GitHub.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+- **`complete_ledgers`** - This field shows which [ledger indexes](basic-data-types.html#ledger-index) your server has complete ledger data for. Healthy servers usually have a single range of recent ledgers, such as `"12133424-12133858"`.
+
+    - If you have a disjoint set of complete ledgers such as `"11845721-12133420,12133424-12133858"`, that could indicate that your server has had intermittent outages or has temporarily fallen out of sync with the rest of the network. The most common causes for this are insufficient disk I/O or network bandwidth.
```
```
< If you have a disjoint set
---
> If you have a disjointed set
```

.. OTOH, isn't a set disjointed by nature in that each element is discrete?

mDuo13 replied:
"disjoint set" is the [correct term](https://en.wikipedia.org/wiki/Disjoint_sets)

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
````diff
@@ -0,0 +1,74 @@
+rippled server_info
+```
+
+The response to this command has a lot of information, which is documented along with the [server_info method][].
````
Maybe include an abbreviated sample of a response?

mDuo13 replied:
I'm not sure how useful it is to include something like this, when the following bullet points contain almost the exact same info:

```
{
   "result" : {
      "info" : {
         "build_version" : "1.1.0-rc3",
         "complete_ledgers" : "12329694-12330647",
         
         ...
         
         "peers" : 11,
         "server_state" : "proposing",
         
         ...
         
      },
      "status" : "success"
   }
}
```

You could also just click the link to see an example with a full response.

What do you think? Is that useful?

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+## Check the server log
+
+While running, `rippled` servers write information to a debug log. The location of the debug log depends on your server's configuration file. The [default configuration](https://github.com/ripple/rippled/blob/master/cfg/rippled-example.cfg#L1139-L1142) writes the server's debug log to the file `/var/log/rippled/debug.log`. If you start the `rippled` service directly (instead of using `systemctl`  ...
```
Maybe simply: 

> ## Check the server log
> 
> The [default configuration](https://github.com/ripple/rippled/blob/master/cfg/rippled-example.cfg#L1139-L1142) writes the server's debug log to `/var/log/rippled/debug.log`. The location can differ depending on your server's configuration file. If you start ...

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+While running, `rippled` servers write information to a debug log. The location of the debug log depends on your server's configuration file. The [default configuration](https://github.com/ripple/rippled/blob/master/cfg/rippled-example.cfg#L1139-L1142) writes the server's debug log to the file `/var/log/rippled/debug.log`. If you start the `rippled` service directly (instead of using `systemctl`  ...
+
+You can control the verbosity of the debug log with the [log_level method][]. The default config file sets the `log_level` to severity "warning" for all categories of log messages. (See the `[rpc_startup]` stanza of the config file for settings.)
```
Maybe switch the order of the first two sentences: 

> The default config file sets the `log_level` to severity "warning" for all categories of log messages. You can control the verbosity of the debug log with the [log_level method][]. (See the `[rpc_startup]` stanza of the config file for settings.)

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+You can control the verbosity of the debug log with the [log_level method][]. The default config file sets the `log_level` to severity "warning" for all categories of log messages. (See the `[rpc_startup]` stanza of the config file for settings.)
+
+It is normal for a `rippled` the server to print many warning-level (`WRN`) messages during startup and a few warning-level messages from time to time later on. You can **safely ignore** most warnings in the first 5 to 15 minutes of server startup.
```
```
< (`WRN`)
---
(`WARN`)  ?
```

mDuo13 replied:
No; the log messages use the prefix `WRN` for warning-level messages.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
```diff
@@ -0,0 +1,169 @@
+
+        ulimit -Hn
+
+    The command prints the hard limit on the number of open files, which should be 10240.
```
Maybe remove "The command ... " and change the first sentence to: 

> 2. Check that the [hard limit of files that can be opened](https://ss64.com/bash/ulimit.html) (at the process level) is now `10240`:

mtrippled replied:
Set those to 65536 each. There's no reason not to, it's how production is.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
```diff
@@ -0,0 +1,169 @@
+
+        systemctl start rippled
+
+4. If `rippled` still does not start, open `/etc/sysctl.conf` and add the following line to the bottom of the file:
```
Maybe: 

> 4. If `rippled` still does not start, open `/etc/sysctl.conf` and append the following kernel level setting:

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
```diff
@@ -0,0 +1,169 @@
+
+Possible solutions:
+
+- Check that `/etc/opt/ripple/rippled.cfg` exists and the `rippled` user has read permissions to the file. (Assuming you use the `rippled` user to run the `rippled` process, and you want to use the default location for the config file.)
```
"Assuming .." is a fragment. Maybe ...

```
< Check that `/etc/opt/ripple/rippled.cfg` exists and the `rippled` user has read permissions to the file. (Assuming you use the `rippled` user to run
---
> Check that `rippled.cfg` exists (the default location is `/etc/opt/ripple/rippled.cfg`) and that the user that runs your `rippled` process (usually `rippled`) has read permissions to it.
```

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
```diff
@@ -0,0 +1,169 @@
+
+- Check that `/etc/opt/ripple/rippled.cfg` exists and the `rippled` user has read permissions to the file. (Assuming you use the `rippled` user to run the `rippled` process, and you want to use the default location for the config file.)
+
+- Create a config file that can be read by the `rippled` user to `$HOME/.config/ripple/rippled.cfg` (where `$HOME` points to the `rippled` user's home directory).
```
```
< Create a config file that can be read by x to y 
---
> Create a config file that can be read by x in y
```

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
````diff
@@ -0,0 +1,169 @@
+rm -r /var/lib/rippled/db
+```
+
+**Tip:** It is generally safe to delete the `rippled` databases, because any individual server can re-download ledger history from other servers in the XRP Ledger network. If you are using [clustering](clustering.html), be sure your servers each have a unique `[node_seed]` configured first; if not, servers may not be recognized as part of the cluster after you restart them.
````
Should this be a caution or warning ... considering the part about unique `node_seed`?

mDuo13 replied:
Better yet, I can just remove that part. The instructions for clustering explicitly tell you to generate and configure a `[node_seed]` so it's a pretty pointless warning to tell people that clustering won't work properly if you delete the databases while not using clustering properly.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
````diff
@@ -0,0 +1,169 @@
+Terminating thread rippled: main: unhandled St13runtime_error 'online_delete must not be less than ledger_history (currently 3000)
+```
+
+The `[ledger_history]` setting represents how many ledgers of history the server should seek to back-fill. The `online_delete` field (in the `[node_db]` stanza) indicates how many ledgers of history to keep when dropping older history. The `online_delete` value must be equal or larger than `[ledger_history]` to prevent the server from deleting historical ledgers that it is also trying to download ...
````
```
< The `online_delete` value must be equal or larger than
---
> The `online_delete` value must be equal to or larger than
```

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
````diff
@@ -0,0 +1,169 @@
+```text
+Loading: "/etc/opt/ripple/rippled.cfg"
+Failed to open '"/etc/opt/ripple/rippled.cfg"'.
+Terminating thread rippled: main: unhandled St13runtime_error 'Can not create "/var/opt/ripple"'
````
I realize that this message comes from the code, but maybe someone can change `can not` to `cannot`.

mDuo13 replied:
¯\_(ツ)_/¯

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
````diff
@@ -0,0 +1,169 @@
+
+```text
+Loading: "/home/rippled/.config/ripple/rippled.cfg"
+Terminating thread rippled: main: unhandled St13runtime_error 'Can not create "/var/lib/rippled/db"'
````
cannot

mDuo13 replied:
Leaving as-is because the message originates in the code

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/understanding-log-messages.md:
```diff
@@ -0,0 +1,139 @@
+A large number of these messages around the same time may indicate a problem, such as:
+
+- Your internet connection to one or more specific peers was cut off
+- Your server may have been overloading the peer with requests, causing it to drop your server
```
Add a period to both sentences.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/understanding-log-messages.md:
````diff
@@ -0,0 +1,139 @@
+## LoadMonitor Job
+
+```text
+2018-Aug-28 22:56:36.180827973 LoadMonitor:WRN Job: gotFetchPack run: 11566ms wait: 0ms
````
Okay, I see .. `WRN` is correct. Hm.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/understanding-log-messages.md:
```diff
@@ -0,0 +1,139 @@
+
+- The server writing the message is having network issues
+- The validator described in the message is having network issues
+- The validator described in the message is behaving maliciously
```
These sentences need periods.

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/diagnosing-problems.md:
```diff
@@ -0,0 +1,74 @@
+
+If your server does not start, see [rippled Server Won't Start](server-wont-start.html) for a list of some possible causes and fixes.
+
+The remainder of this document suggests steps for diagnosing the problem if your server starts successfully.
```
Can a server start unsuccessfully? Maybe: 

```
< The remainder of this document suggests steps for diagnosing the problem if your server starts successfully.
---
> This document covers problems you may encounter when your server is up and running. 
```

gbarr01 commented on content/tutorials/manage-the-rippled-server/troubleshooting/server-wont-start.md:
```diff
@@ -0,0 +1,169 @@
+# rippled Server Won't Start
+
+This page explains possible reasons the `rippled` server does not start successfully, and how to fix them.
```
```
< This page explains possible reasons the `rippled` server does not start successfully, and how
---
> This page explains possible reasons the `rippled` server does not start and how
```

mDuo13 replied:
I mean, debatably, the server usually _starts_ but doesn't always get through the entire startup successfully. ;) But point taken.

gbarr01 reviewed (approved):
LGTM. Lots of nits but, overall, great docs. Useful.

mtrippled reviewed (commented):
I left a few comments already without realizing to put them into the review, but I'm sure you'll find them @mDuo13

mDuo13 committed:
Troubleshooting - edits per reviews

mDuo13 commented:
I think I've addressed all the reviewer feedback so far, but still waiting for info on https://github.com/ripple/ripple-dev-portal/pull/453/files#diff-8f38126fe6f6261393f95ebbf7022aeaR49

mDuo13 committed:
Troubleshooting: Clarify 'no hash for fetch pack' error

intelliot reviewed (approved):
Great work collecting this info about the log messages! Lots of stuff I didn't know.

I think it would be handy to add some info about switching to NuDB, especially since this has solved problems for me in the past. (And the default cfg uses RocksDB)

mtrippled reviewed (approved):
:+1:
